### PR TITLE
feat: add fleet,kibana,synthetics,heartbeat e2e tests

### PR DESCRIPTION
**Pull request by dominiqueclarke**
Relates to #360

Adds e2e tests for the Synthetics integration

### Why is this important?
These tests provide a way for us to have confidence in the integration between Kibana, elastic-agent, heartbeat and the synthetics agents. 

We currently have functional tests for the Synthetics integration in Kibana, but these tests only test the expected package policy output, and do not run against elastic-agent.

### Why is it in this repo?

While synthetics is available in Kibana's functional test runner, fleet server and elastic-agent is not. Running `elastic-package` gives us the ability the easily bootstrap the whole stack, including Kibana, Elasticsearch, Fleet-server, elastic-agent, and package-registry

### CODE REVIEW

File: __tests__/e2e/scripts/setup.sh
Comment (dominiqueclarke) [inline __tests__/e2e/scripts/setup.sh:5]: The container CI is run in would need to have docker and go installed.

File: __tests__/e2e/scripts/setup.sh
Comment (dominiqueclarke) [inline __tests__/e2e/scripts/setup.sh:13]: We would need an environment variable provided in the container to pass through the version we'd like to test.

File: __tests__/e2e/synthetics.journey.ts
Comment (dominiqueclarke) [inline __tests__/e2e/synthetics.journey.ts:121]: ICMP tests are currently disabled, as icmp requires `elastic-agent` to be run as root.

File: __tests__/e2e/scripts/setup.sh
Comment (vigneshshanmugam) [reply __tests__/e2e/scripts/setup.sh:13]: Do you need help from the robots team on this?

File: __tests__/e2e/synthetics.journey.ts
Comment (andrewvc) [inline __tests__/e2e/synthetics.journey.ts:47]: Can we comment any uses of `waitForTimeout`? It's good to know in the future so we know what exactly we're waiting for. Ideally we would wait instead for an element.

**Review by andrewvc — COMMENTED** [review]
Generally looking good, left a couple nits

File: __tests__/e2e/synthetics.journey.ts
Comment (dominiqueclarke) [reply __tests__/e2e/synthetics.journey.ts]: Great point. @andrewvc I've removed all extraneous uses of `waitForTimeout` and commented the one remaining use. This is ready for a second review.

File: __tests__/e2e/synthetics.journey.ts
Comment (vigneshshanmugam) [inline __tests__/e2e/synthetics.journey.ts:48]: If you want to ensure the page is fully loaded as a result of the click event
```
const nextPage = await Promise.all([
   await page.click(),
   await page.waitForUrl("") // new page
]);
```

would this help here?

File: __tests__/e2e/scripts/test_integration.sh
Comment (andrewvc) [inline __tests__/e2e/scripts/test_integration.sh:6]:
```suggestion
npx @elastic/synthetics synthetics.journey.ts $SYNTHETICS_E2E_ARGS
```

Can we make no-headless optional? Otherwise this doesn't work on headless environments (like my test linux box0

**Review by andrewvc — APPROVED** [review]
LGTM after fixing one minor nit. Great work!

File: __tests__/e2e/package.json
Comment (vigneshshanmugam) [inline __tests__/e2e/package.json:9]: the ideal way in NPM scripts is to use colons `setup:integration`, `test:integration` etc.

File: __tests__/e2e/package.json
Comment (vigneshshanmugam) [inline __tests__/e2e/package.json:5]: remove main from this package.json

File: __tests__/e2e/scripts/setup_integration.sh
Comment (vigneshshanmugam) [inline __tests__/e2e/scripts/setup_integration.sh:10]: Is this a catch flag that takes stack down if the stack was up previously from other tests? 

What happens if one runs setup_integration multiple times? 

Its better to group the setup and test integration together and RIP the stack at one place. The current `setup` and `test` feels a bit out of the way as setup has to be run before the test.

File: __tests__/e2e/scripts/test_integration.sh
Comment (vigneshshanmugam) [inline __tests__/e2e/scripts/test_integration.sh:9]: As discussed in the previous comment. 

```
1. Install elastic-package
2. Boot up the stack with given version
3. Run synthetics journeys
4. Bring the stack down
```

File: __tests__/e2e/synthetics.journey.ts
Comment (vigneshshanmugam) [inline __tests__/e2e/synthetics.journey.ts:58]: cant we search for something better here? Like `text=Monitors` or some headlines with explicit title?

**Review by vigneshshanmugam — COMMENTED** [review]
Great work @dominiqueclarke

File: __tests__/e2e/scripts/setup_integration.sh
Comment (dominiqueclarke) [reply __tests__/e2e/scripts/setup_integration.sh:10]: `test:integration` actually runs both the test and setup script. This is also how the existing tests work.

This elastic-package stack down is a bit of a fail safe. We actually do run this command after the tests finish as well, but this ensures that you do not have an elastic-package stack up when running locally. This is most helpful for local development, and is pretty irrelevant in the ci context.

File: __tests__/e2e/synthetics.journey.ts
Comment (dominiqueclarke) [reply __tests__/e2e/synthetics.journey.ts:58]: The h1 is actually a different between different versions, due to the recent "No data" screen. So this accounts for all of the variations. I could also do `await page.waitForUrl("")` instead if you prefer.

Comment (vigneshshanmugam) [reply __tests__/e2e/synthetics.journey.ts:58]: Hmmm gotcha, if its going to be a explicit URL then it would be better. I have found it hard when trying to debug these tests before. So more verbose would be always preferable.

File: __tests__/e2e/scripts/ci_integration.sh
Comment (vigneshshanmugam) [inline __tests__/e2e/scripts/ci_integration.sh:20]: I dont think this is correct, this is a disruptive command and should be used with caution. We should only prune stuff that was started by running the test.

File: __tests__/e2e/scripts/setup_integration.sh
Comment (vigneshshanmugam) [reply __tests__/e2e/scripts/setup_integration.sh:10]: I actually had multiple problems with the current setup of having setup step and test step where in theory we could merge both and have a single script that effectively handles starting and killing containers on failure. But this is not the scope for this PR, so lets keep whatever way you prefer,

**Review by v1v — CHANGES_REQUESTED** [review]
I'm now working and contributing to this PR in order to enable the e2e tests to run in the CI.

Please do not merge it yet

File: .ci/e2e-synthetics.groovy
Comment (v1v) [inline .ci/e2e-synthetics.groovy:46]: This is the new task to be run, can you confirm if this is the corrected one?

File: .ci/e2e-synthetics.groovy
Comment (v1v) [inline .ci/e2e-synthetics.groovy:96]: TBR: see https://github.com/elastic/apm-pipeline-library/pull/1402

**Comment by v1v** [comment]
I managed to configure this new e2e testing in the CI and it failed with the below error:

```
[2021-11-22T15:09:01.230Z] npm WARN rollback Rolling back ts-node@10.4.0 failed (this is probably harmless): /var/lib/jenkins/workspace/nt-rum_e2e-synthetics-mbp_PR-393/src/github.com/elastic/synthetics/node_modules/ts-node is not a child of /var/lib/jenkins/workspace/nt-rum_e2e-synthetics-mbp_PR-393/src/github.com/elastic/synthetics/__tests__/e2e
[2021-11-22T15:09:01.230Z] npm ERR! code ENOENT
[2021-11-22T15:09:01.230Z] npm ERR! syscall chmod
[2021-11-22T15:09:01.230Z] npm ERR! path /var/lib/jenkins/workspace/nt-rum_e2e-synthetics-mbp_PR-393/src/github.com/elastic/synthetics/__tests__/e2e/node_modules/@elastic/synthetics/dist/cli.js
[2021-11-22T15:09:01.230Z] npm ERR! errno -2
[2021-11-22T15:09:01.230Z] npm ERR! enoent ENOENT: no such file or directory, chmod '/var/lib/jenkins/workspace/nt-rum_e2e-synthetics-mbp_PR-393/src/github.com/elastic/synthetics/__tests__/e2e/node_modules/@elastic/synthetics/dist/cli.js'
[2021-11-22T15:09:01.230Z] npm ERR! enoent This is related to npm not being able to find a file.
[2021-11-22T15:09:01.230Z] npm ERR! enoent 
[2021-11-22T15:09:01.230Z] 
[2021-11-22T15:09:01.230Z] npm ERR! A complete log of this run can be found in:
[2021-11-22T15:09:01.230Z] npm ERR!     /var/lib/jenkins/workspace/nt-rum_e2e-synthetics-mbp_PR-393/.npm/_logs/2021-11-22T15_09_01_193Z-debug.log
[2021-11-22T15:09:01.230Z] npm ERR! code ELIFECYCLE
[2021-11-22T15:09:01.230Z] npm ERR! errno 254
[2021-11-22T15:09:01.230Z] npm ERR! e2e-synthetics@ test:ci_integration_all: `npm i && bash scripts/test_versions.sh ci`
[2021-11-22T15:09:01.230Z] npm ERR! Exit status 254
[2021-11-22T15:09:01.230Z] npm ERR! 
[2021-11-22T15:09:01.230Z] npm ERR! Failed at the e2e-synthetics@ test:ci_integration_all script.
[2021-11-22T15:09:01.230Z] npm ERR! This is probably not a problem with npm. There is likely additional logging output above.
```

@dominiqueclarke ,  would you mind helping on this?

File: .ci/e2e-synthetics.groovy
Comment (v1v) [inline .ci/e2e-synthetics.groovy:43]: 1) install node as required

File: .ci/e2e-synthetics.groovy
Comment (v1v) [inline .ci/e2e-synthetics.groovy:44]: 2. install go with the required package

File: .ci/e2e-synthetics.groovy
Comment (v1v) [reply .ci/e2e-synthetics.groovy:46]: 3. run the e2e tests

**Comment by dominiqueclarke** [comment]
> go get

Hi @v1v 

Looks like there are two issues.

1. You'll need to run `go install github.com/elastic/elastic-package@latest` instead of `go get github.com/elastic/elastic-package` due to `go get` being deprecated in 1.17 https://go.dev/doc/go-get-install-deprecation. This step is actually taken care of in our `setup_integration` script [here](https://github.com/elastic/synthetics/pull/393/files#diff-de7cdf97c6f5ed09a814caaf876d949e9d71344239c216694340d4f47f296292R5), but if we need to move it outside of the bash script, we'll still need to use `go install` instead of `go get`
2. Looks like in my README, I neglected the `npm install` step. I've updated the README. You'll need to run `npm install` inside the __tests__/e2e directory

File: .ci/e2e-synthetics.groovy
Comment (v1v) [reply .ci/e2e-synthetics.groovy:44]:
```suggestion
            withGoEnv(){
```

File: .ci/e2e-synthetics.groovy
Comment (v1v) [reply .ci/e2e-synthetics.groovy:46]:
```suggestion
                sh(label: 'npm install', script: 'npm install')
                sh(label: 'run e2e tests', script: 'npm run test:ci_integration_all')
```

**Comment by v1v** [comment]
> You'll need to run npm install inside the tests/e2e directory

@dominiqueclarke , I'm afraid `npm install` fails with a similar error

```
[2021-11-22T22:24:16.323Z] npm ERR! code ENOENT
[2021-11-22T22:24:16.323Z] npm ERR! syscall chmod
[2021-11-22T22:24:16.323Z] npm ERR! path /var/lib/jenkins/workspace/nt-rum_e2e-synthetics-mbp_PR-393/src/github.com/elastic/synthetics/__tests__/e2e/node_modules/@elastic/synthetics/dist/cli.js
[2021-11-22T22:24:16.323Z] npm ERR! errno -2
[2021-11-22T22:24:16.323Z] npm ERR! enoent ENOENT: no such file or directory, chmod '/var/lib/jenkins/workspace/nt-rum_e2e-synthetics-mbp_PR-393/src/github.com/elastic/synthetics/__tests__/e2e/node_modules/@elastic/synthetics/dist/cli.js'
[2021-11-22T22:24:16.323Z] npm ERR! enoent This is related to npm not being able to find a file.
[2021-11-22T22:24:16.323Z] npm ERR! enoent 
[2021-11-22T22:24:16.323Z] 
[2021-11-22T22:24:16.323Z] npm ERR! A complete log of this run can be found in:
[2021-11-22T22:24:16.323Z] npm ERR!     /var/lib/jenkins/workspace/nt-rum_e2e-synthetics-mbp_PR-393/.npm/_logs/2021-11-22T22_24_15_042Z-debug.log
```

See https://github.com/elastic/synthetics/pull/393/commits/14192bb417668a9e66f904d1742d6d82e4d9761b

**Comment by v1v** [comment]
It moved further with my recent changes but:

```
[2021-11-22T23:10:20.097Z] ./scripts/ci_integration.sh: 2: set: Illegal option -o pipefail
[2021-11-22T23:10:20.097Z] npm ERR! code ELIFECYCLE
[2021-11-22T23:10:20.097Z] npm ERR! errno 2
[2021-11-22T23:10:20.097Z] npm ERR! e2e-synthetics@ test:ci_integration_all: `npm i && bash scripts/test_versions.sh ci`
[2021-11-22T23:10:20.097Z] npm ERR! Exit status 2
[2021-11-22T23:10:20.097Z] npm ERR! 
[2021-11-22T23:10:20.097Z] npm ERR! Failed at the e2e-synthetics@ test:ci_integration_all script.
[2021-11-22T23:10:20.097Z] npm ERR! This is probably not a problem with npm. There is likely additional logging output above.
[2021-11-22T23:10:20.359Z] 
[2021-11-22T23:10:20.359Z] npm ERR! A complete log of this run can be found in:
[2021-11-22T23:10:20.359Z] npm ERR!     /var/lib/jenkins/workspace/nt-rum_e2e-synthetics-mbp_PR-393/.npm/_logs/2021-11-22T23_10_20_024Z-debug.log
script returned exit code 2
```

File: .ci/e2e-synthetics.groovy
Comment (v1v) [inline .ci/e2e-synthetics.groovy:59]:
```suggestion
              sh(label: 'run e2e tests', script: '''#!/bin/bash
                npm run test:ci_integration_all
              ''')
```

File: __tests__/e2e/scripts/ci_integration.sh
Comment (v1v) [inline __tests__/e2e/scripts/ci_integration.sh:2]: As far as I see `o pipefail` is not required since `|` (pipes) are not present

File: __tests__/e2e/scripts/test_integration.sh
Comment (v1v) [inline __tests__/e2e/scripts/test_integration.sh:4]: Typo? Missing `-o` in order to use `pipefail` in addition no pipes are in this script, so do you think it can be simplifier with the below suggestion?

```suggestion
set -e
```

**Comment by dominiqueclarke** [comment]
> https://golang.org/doc/go-get-install-deprecation

The funny thing is, I thought about it further after I posted and realized I was wrong because `npm I` is included in the scripts. I began investigating further with the intent to update the comment. I didn't expect you to be working at 11:00PM. Thank you for further investigating. I'll review all your comments later tonight.

**Comment by dominiqueclarke** [comment]
> IIUC, the `dist` folder in the root of the project doesn't exist, since in order to run the `e2e` is required to build the project beforehand. I'm guessing and trying another changes in [e5547ba](https://github.com/elastic/synthetics/commit/e5547ba70202d606272a9c6942790f8a6c3c4714) and [16e92a0](https://github.com/elastic/synthetics/commit/16e92a0b81becd8e7f0d1ff10132b039292f2b3e) . Let's see if that solves the existing issue. Unfortunately, I could not find any references about this, but for the error message, it seems that could be the issue, or maybe not...

@v1v You are right, the synthetics agent needs to be built first. We can add that as part of the script or you can add building the agent first into the workflow.

**Comment by v1v** [comment]
I triggered again to test with the latest changes and it failed in the same script but something else this time

```
[2021-11-23T09:26:43.801Z] ./scripts/ci_integration.sh: 1: ./scripts/ci_integration.sh: Bad substitution
[2021-11-23T09:26:43.801Z] npm ERR! code ELIFECYCLE
[2021-11-23T09:26:43.801Z] npm ERR! errno 2
[2021-11-23T09:26:43.801Z] npm ERR! e2e-synthetics@ test:ci_integration_all: `npm i && bash scripts/test_versions.sh ci`
[2021-11-23T09:26:43.801Z] npm ERR! Exit status 2
```

File: __tests__/e2e/scripts/test_integration.sh
Comment (v1v) [inline __tests__/e2e/scripts/test_integration.sh:4]: The shebang should be added at the beginning of the script, so this might cause some issues.

File: __tests__/e2e/scripts/ci_integration.sh
Comment (v1v) [inline __tests__/e2e/scripts/ci_integration.sh:4]: The shebang should be added at the beginning of the script, so this might cause some issues

**Comment by v1v** [comment]
I even tried locally and faced some issues:

```bash
$ gvm 1.17.3
export GOROOT="/Users/vmartinez/.gvm/versions/go1.17.3.darwin.amd64"
export PATH="/Users/vmartinez/.gvm/versions/go1.17.3.darwin.amd64/bin:$PATH"
$ nvm use
Found '/Users/vmartinez/work/src/github.com/elastic/synthetics/.nvmrc' with version <14.17.5>
Now using node v14.17.5 (npm v6.14.14)
$ rm -rf node_modules 
$ npm install

> husky@4.3.8 install /Users/vmartinez/work/src/github.com/elastic/synthetics/node_modules/husky
> node husky install

husky > Setting up git hooks
husky > Done

> playwright-chromium@1.14.0 install /Users/vmartinez/work/src/github.com/elastic/synthetics/node_modules/playwright-chromium
> node install.js

Downloading Playwright build of chromium v907428 - 117.5 Mb [====================] 100% 0.0s 
Playwright build of chromium v907428 downloaded to /Users/vmartinez/Library/Caches/ms-playwright/chromium-907428
Downloading Playwright build of ffmpeg v1005 - 1.3 Mb [====================] 100% 0.0s 
Playwright build of ffmpeg v1005 downloaded to /Users/vmartinez/Library/Caches/ms-playwright/ffmpeg-1005

> sharp@0.28.3 install /Users/vmartinez/work/src/github.com/elastic/synthetics/node_modules/sharp
> (node install/libvips && node install/dll-copy && prebuild-install) || (node install/can-compile && node-gyp rebuild && node install/dll-copy)

sharp: Downloading https://github.com/lovell/sharp-libvips/releases/download/v8.10.6/libvips-8.10.6-darwin-x64.tar.br

> husky@4.3.8 postinstall /Users/vmartinez/work/src/github.com/elastic/synthetics/node_modules/husky
> opencollective-postinstall || exit 0

Thank you for using husky!
If you rely on this package, please consider supporting our open collective:
> https://opencollective.com/husky/donate

npm WARN prepublish-on-install As of npm@5, `prepublish` scripts are deprecated.
npm WARN prepublish-on-install Use `prepare` for build steps and `prepublishOnly` for upload-only.
npm WARN prepublish-on-install See the deprecation note in `npm help scripts` for more information.

> @elastic/synthetics@1.0.0-beta.17 prepublish /Users/vmartinez/work/src/github.com/elastic/synthetics
> npm run clean && npm run build


> @elastic/synthetics@1.0.0-beta.17 clean /Users/vmartinez/work/src/github.com/elastic/synthetics
> rimraf dist


> @elastic/synthetics@1.0.0-beta.17 build /Users/vmartinez/work/src/github.com/elastic/synthetics
> tsc

npm WARN @elastic/synthetics@1.0.0-beta.17 No repository field.

added 811 packages from 667 contributors and audited 814 packages in 38.488s

72 packages are looking for funding
  run `npm fund` for details

found 27 vulnerabilities (26 moderate, 1 high)
  run `npm audit fix` to fix them, or `npm audit` for details
npm run build

> @elastic/synthetics@1.0.0-beta.17 build /Users/vmartinez/work/src/github.com/elastic/synthetics
> tsc
$ cd __tests__/e2e 
$ npm install
audited 184 packages in 2.275s

73 packages are looking for funding
  run `npm fund` for details

found 3 moderate severity vulnerabilities
  run `npm audit fix` to fix them, or `npm audit` for details

$ npm run test:ci_integration_all

> e2e-synthetics@ test:ci_integration_all /Users/vmartinez/work/src/github.com/elastic/synthetics/__tests__/e2e
> npm i && bash scripts/test_versions.sh ci

audited 184 packages in 2.22s

73 packages are looking for funding
  run `npm fund` for details

found 3 moderate severity vulnerabilities
  run `npm audit fix` to fix them, or `npm audit` for details
./scripts/setup_integration.sh: line 7: elastic-package: command not found
./scripts/setup_integration.sh: line 10: elastic-package: command not found
./scripts/setup_integration.sh: line 7: elastic-package: command not found
./scripts/setup_integration.sh: line 10: elastic-package: command not found
./scripts/setup_integration.sh: line 7: elastic-package: command not found
./scripts/setup_integration.sh: line 10: elastic-package: command not found
npm ERR! code ELIFECYCLE
npm ERR! syscall spawn
npm ERR! file sh
npm ERR! errno ENOENT
npm ERR! e2e-synthetics@ test:ci_integration_all: `npm i && bash scripts/test_versions.sh ci`
npm ERR! spawn ENOENT
npm ERR! 
npm ERR! Failed at the e2e-synthetics@ test:ci_integration_all script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

npm ERR! A complete log of this run can be found in:
npm ERR!     /Users/vmartinez/.npm/_logs/2021-11-23T09_47_14_122Z-debug.log
```


Somehow the elastic-package is not installed as part of the `npm run test:ci_integration_all`, can you please clarify if that's the case? Or did I miss anything in my local development?

**Comment by dominiqueclarke** [comment]
> ERROR: for elastic-agent_is_ready  Container "6d624458d7ca" is unhealthy.

Unhealthy containers are most often caused for 3 different reasons
1. Docker does not have enough resources to start up the elastic stack. Marcin Tojek recommends these settings
```
CPUs: 12, Memory: 24GB, Swap: 4GB, Disk: 60GB
```
2. Old images are not pruned appropriately. The images brought up by elastic-package should be removed on each step as part of the script.
3. Snapshots are broken

Those are the three most common issues with containers. If the situation isn't resolved by adjusting docker images or ensuring that cached images are deleted, and we're sure that snapshot isn't broken, we'll have to continue down the debug path. 

As far as the warning of the commit hash, that is expected and harmless warn when using elastic-package with `go install` instead of building from source. I suppose we could build it from source, but I thought `go install` would be easier.

**Comment by v1v** [comment]
Thanks for the clarification. 

[build10](https://apm-ci.elastic.co/blue/organizations/jenkins/apm-agent-rum%2Fe2e-synthetics-mbp/detail/PR-393/10/pipeline) is now the one that I just manually triggered.

In order to run the e2e pipeline, since it was required to run on daily basis, any changes in this PR will not trigger a build automatically. Therefore you can login in the Jenkins UI and force a build manually:

- https://apm-ci.elastic.co/job/apm-agent-rum/job/e2e-synthetics-mbp/job/PR-393/10/


The error I saw was, can you please verify if the script supports the existing bash version 4?
And add some debug traces in the script to know what it does if possible?

```
[2021-11-23T09:26:43.801Z] ./scripts/ci_integration.sh: 1: ./scripts/ci_integration.sh: Bad substitution
```

Please poke me if you need further assistance.

**Comment by dominiqueclarke** [comment]
> Thanks for the clarification.
> 
> [build10](https://apm-ci.elastic.co/blue/organizations/jenkins/apm-agent-rum%2Fe2e-synthetics-mbp/detail/PR-393/10/pipeline) is now the one that I just manually triggered.
> 
> In order to run the e2e pipeline, since it was required to run on daily basis, any changes in this PR will not trigger a build automatically. Therefore you can login in the Jenkins UI and force a build manually:
> 
> * https://apm-ci.elastic.co/job/apm-agent-rum/job/e2e-synthetics-mbp/job/PR-393/10/
> 
> The error I saw was, can you please verify if the script supports the existing bash version 4? And add some debug traces in the script to know what it does if possible?
> 
> ```
> [2021-11-23T09:26:43.801Z] ./scripts/ci_integration.sh: 1: ./scripts/ci_integration.sh: Bad substitution
> ```
> 
> Please poke me if you need further assistance.

@v1v This has been resolved. You can see the passing ci runs here https://apm-ci.elastic.co/job/apm-agent-rum/job/e2e-synthetics-mbp/view/change-requests/job/PR-393/

If there's nothing else needed from robots, could you unblock this PR?

File: .ci/e2e-synthetics.groovy
Comment (v1v) [inline .ci/e2e-synthetics.groovy:101]:
```suggestion
```

**Review by v1v — APPROVED** [review]
💯 Thanks

**Pull request merged**
